ksql-engine/src/main/java/io/confluent/ksql/function/udaf/average/AverageUdaf.java

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright 2018 Confluent Inc.
+ *
+ * Licensed under the Confluent Community License (the "License"); you may not use
+ * this file except in compliance with the License.  You may obtain a copy of the
+ * License at
+ *
+ * http://www.confluent.io/confluent-community-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.function.udaf.average;
+
+import io.confluent.ksql.function.udaf.TableUdaf;
+import io.confluent.ksql.function.udaf.UdafDescription;
+import io.confluent.ksql.function.udaf.UdafFactory;
+import org.apache.kafka.connect.data.Schema;
+import org.apache.kafka.connect.data.SchemaBuilder;
+import org.apache.kafka.connect.data.Struct;
+
+@UdafDescription(name = "avg",
+    description = "Returns the average value of the column computed as the sum divided by the"
+        + "count. Applicable only to numeric types.")
+public final class AverageUdaf {
+
+  private static final String COUNT = "COUNT";
+  private static final String SUM = "SUM";
+  private static final Schema STRUCT_LONG = SchemaBuilder.struct().optional()
+      .field(SUM, Schema.OPTIONAL_INT64_SCHEMA)
+      .field(COUNT, Schema.OPTIONAL_INT64_SCHEMA)
+      .build();
+  private static final Schema STRUCT_DOUBLE = SchemaBuilder.struct().optional()
+      .field(SUM, Schema.OPTIONAL_FLOAT64_SCHEMA)
+      .field(COUNT, Schema.OPTIONAL_INT64_SCHEMA)
+      .build();
+
+
+  private AverageUdaf() {
+  }
+
+
+  @UdafFactory(description = "Compute average of column with type Long.",
+      aggregateSchema = "STRUCT<SUM bigint, COUNT bigint>")
+  public static TableUdaf<Long, Struct, Double> averageLong() {
+
+    return new TableUdaf<Long, Struct, Double>() {
+
+      @Override
+      public Struct initialize() {
+        return new Struct(STRUCT_LONG).put(SUM, 0L).put(COUNT, 0L);
+      }
+
+      @Override
+      public Struct aggregate(final Long newValue,
+                              final Struct aggregate) {
+
+        if (newValue == null) {
+          return aggregate;
+        }
+        return new Struct(STRUCT_LONG)
+            .put(SUM, aggregate.getInt64(SUM) + newValue)
```

Comment:
instead of duplicating this `TableUdaf`, you can generate a tableUdaf by passing in the proper getter method (e.g. `private static <I> TableUdaf<I, Struct, Double> getAverageUdaf(final Function<Struct, I> getter)`